### ISSUE TITLE
Firefox reported SEC_ERROR_REUSED_ISSUER_AND_SERIAL for MITM

### ISSUE THREAD

**Issue by ShichangXu** [elazarl/goproxy#311]
I run examples/goproxy-eavesdropper/main.go to perform MITM on HTTPS connections.
I installed the certificate in firefox and open web pages through goproxy.

The webpages are not rendered properly. I use the developer tools in firefox to examine the network activities and found that ~50% of objects are not succeeded downloaded. Firefox reported SEC_ERROR_REUSED_ISSUER_AND_SERIAL for those failed objects. This seems to be issue with the certificates.

I use wireshark to capture the generated certificates from goproxy. It seems that goproxy can generates different certificates even for the same domain. This causes issue for firefox. I guess that firefox will check and make sure the certificate for each domain is constant.

I checked the goproxy code. In signer.signHost(), it calls x509.CreateCertificate(), which eventually calls randutil.MaybeReadByte(). This function generates some randomness and reads a single byte from r with ~50% probability. It seems that this is causing goproxy to generate different certificates for the same domain.

I implemented a quick fix of the code and let goproxy to cache the generated certificate for each domain. This seems to fix this issue.

I am not familiar with Go language that much and not very confident what I found is correct. Could you help me confirm and fix this? Thanks.

**Comment by anhhh11** [comment]
Cached in the go dictionary not enough; when you restart your goproxy application it sometime will appear SEC_ERROR_REUSED_ISSUER_AND_SERIAL error again. I've implemented cached to when open app read to the dictionary for the first time from file; the next request will read from the dictionary; this solution seem very stable not seeing SEC_ERROR_REUSED_ISSUER_AND_SERIAL again.

**Comment by ShichangXu** [comment]
Thanks for the quick response. This solution makes sense.

However, after I changed the code, I cleared the cache and restarted firefox, some webpages are still not rendered properly. For example, all images on Amazon are not shown, as the domain "images-na.ssl-images-amazon.com" still reports SEC_ERROR_REUSED_ISSUER_AND_SERIAL.

I am wondering what is the fundamental cause of this issue. Are there other reasons why this issue is triggered.
